services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/document/DefaultEnhancedDocument.java

```diff
@@ -0,0 +1,559 @@
+    }
+
+    /**
+     * Keeping the backward compatibility with older version of sdk where 0 and 1 are treated a true and false respectively.
+     */
+    private Boolean getBool(String attributeName) {
+        Object object = get(attributeName);
+        if (object instanceof Boolean) {
+            return (Boolean) object;
+        }
+        if (object instanceof String || object instanceof SdkNumber) {
+            if ("1".equals(object.toString())) {
+                return true;
+            }
+            if ("0".equals(object.toString())) {
+                return false;
+            }
+            return Boolean.valueOf((String) object);
+        }
+        throw new IllegalStateException("Value of attribute " + attributeName + " of type " + getTypeOf(attributeName)
+                                        + " cannot be converted into a Boolean value.");
+    }
+
+    @Override
+    public Object get(String attributeName) {
+        AttributeValue attributeValue = attributeValueMap.get(attributeName);
+        if (attributeValue == null) {
+            return null;
+        }
+        return convertAttributeValueToObject(attributeValue);
+    }
+
+    @Override
+    public EnhancedType<?> getTypeOf(String attributeName) {
+        Object attributeValue = get(attributeName);
+        return attributeValue != null ? EnhancedType.of(attributeValue.getClass()) : null;
+    }
+
+    @Override
+    public Map<String, Object> asMap() {
+        Map<String, Object> result = new LinkedHashMap<>();
+        attributeValueMap.forEach((s, attributeValue) -> result.put(s, toSimpleValue(attributeValue)));
+        return result;
+    }
+
+    @Override
+    public String toJson() {
+        AttributeValue jsonMap = AttributeValue.fromM(attributeValueMap);
+        JsonItemAttributeConverter jsonItemAttributeConverter = JsonItemAttributeConverter.create();
+        JsonNode jsonNode = jsonItemAttributeConverter.transformTo(jsonMap);
+        Document document = jsonNode.visit(new DocumentUnmarshaller());
+        return document.toString();
+    }
+
+    @Override
+    public String toJsonPretty() {
+        return null;
+    }
+
+    public static class DefaultBuilder implements EnhancedDocument.Builder {
+
+        Map<String, Object> attributeValueMap = new LinkedHashMap<>();
+
+        List<AttributeConverterProvider> attributeConverterProviders = new ArrayList<>();
+
+        private DefaultBuilder() {
+        }
+
+        @Override
+        public Builder add(String attributeName, Object value) {
+            attributeValueMap.put(attributeName, value);
+            return this;
+        }
+
+        @Override
+        public Builder addString(String attributeName, String value) {
+            Validate.isTrue(!StringUtils.isEmpty(attributeName), "attributeName cannot empty or null");
+            if (!isNullValueAdded(attributeName, value)) {
+                attributeValueMap.put(attributeName, AttributeValue.fromS(value));
+            }
```

Comment:
We definitely should be using attribute value converters for these methods as well.